### PR TITLE
Refactored metadata linting

### LINKED ISSUE #2006: `tmt lint` fails with fmf path not in root
Here is a minimum example:
```console
$ tree -a
.
└── a
    ├── .fmf
    │   └── version
    └── plan.fmf
$ tmt --root a lint
/plan
pass correct attributes are used
fail No tree found in repo 'None', missing an '.fmf' directory?
```
with `plan.fmf`:
```yaml
summary:
  simple example
discover:
  how: fmf
  filter: "tag: something"
execute:
  how: tmt
```

`tmt --root a run` works fine, but I did not confirm that others work.

### PR DESCRIPTION
**Pull request by happz**
The existing user-facing API did not change, we still have the CLI commands we're used to, `tmt [test|plan|story] lint` and `tmt lint`. internally, things are not different, with linters split into categories, automagical discovery, formalized internal APIs, and new CLI look (`--list-checks`, `--failed-only`, `--enable-checks` & brand new output).

```
$ tmt story lint /spec/plans/import
/spec/plans/import
pass C000 fmf node passes schema validation
pass C001 summary key is set
pass C002 summary is reasonably long
pass S001 correct key are used
fail S002 story is required
```

Fixes #2006

### CODE REVIEW

**Comment by thrix** [comment]
@happz while running it manually, this looks weird:

```
warn C001 summary key is missing, it is useful for quick inspection
skip C002 summary key is missing, it is useful for quick inspection
```

**Comment by thrix** [comment]
@happz also we will need to add some documentation about this, describing all the available checks?

File: tests/lint/test.sh [hunk 1]
```diff
@@ -33,18 +33,33 @@ rlJournalStart
...
         # linting test
-        rlAssertGrep "pass test script must be defined" $rlRun_LOG
-        rlAssertGrep "pass directory path must be absolute" $rlRun_LOG
-        rlAssertGrep "pass directory path must exist" $rlRun_LOG
-        rlAssertGrep "warn summary is very useful for quick inspection"   \
-            $rlRun_LOG
-        rlAssertGrep "fail unknown attribute 'summarrry' is used" $rlRun_LOG
+        rlAssertGrep "warn C000 fmf node failed schema validation" $rlRun_LOG
+        rlAssertGrep "warn C001 summary key is missing, it is useful for quick inspection" $rlRun_LOG
+        rlAssertGrep "skip C002 summary key is missing, it is useful for quick inspection" $rlRun_LOG
```
Comment (thrix) [inline tests/lint/test.sh:44]: this warn/skip combo looks confusing to me, why it is designed this way? also the checks look the same, at least the text is the same.

File: tmt/lint.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
+# coding: utf-8
+
+"""
+Metadata linting.
+
+Internal APIs, classes and shared functionality and helpers for linting of
```
Comment (thrix) [inline tmt/lint.py:6]:
```suggestion
Internal APIs, classes, shared functionality and helpers for linting of
```

**Review by thrix — CHANGES_REQUESTED** [review]
Some small comments, otherwise the code looks clean

File: tmt/lint.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    LinterOutcome.FIXED: 'green'
+    }
+
+
+#: Info on how a linter decided: linter itself, its outcome & the message.
+LinterRuling = Tuple['Linter', LinterOutcome, str]
+#: A return value type of a single linter.
+LinterReturn = Generator[Tuple[LinterOutcome, str], None, None]
+#: A linter itself, a callable method.
+LinterCallback = Callable[['Lintable'], LinterReturn]
```
Comment (janhavlin) [inline tmt/lint.py:125]: Variables used only for type annotations usually have `Type` suffix?
```suggestion
LinterRulingType = Tuple['Linter', LinterOutcome, str]
#: A return value type of a single linter.
LinterReturnType = Generator[Tuple[LinterOutcome, str], None, None]
#: A linter itself, a callable method.
LinterCallbackType = Callable[['Lintable'], LinterReturnType]
```

File: tests/lint/test.sh (hunk 1 above)
Comment (happz) [reply tests/lint/test.sh:44]: We can improve those. There are two distinct checks:

```shell
$ tmt test lint --list-checks
Linters available for tests
...
C001: summary key should be set, it is very useful for quick inspection
C002: summary should not exceed 50 characters
```

When the summary key is not set at all - reported as a warning, `warn C001 summary key is missing`, the second test is skipped. I can change it to a warning, but it sort of makes sense to me, as there's no `summary` to check for length.

I'm open to suggestions :)

File: tmt/lint.py (hunk 3 above)
Comment (happz) [reply tmt/lint.py:125]: That's a question. Recently I tend towards not using the `Type` suffix, as the "type" is sort of already implied - `-> Foo` means "the function returns a value of type `Foo`", therefore I started to consider `FooType` as superfluous.

**Comment by happz** [comment]
> @happz also we will need to add some documentation about this, describing all the available checks?

@thrix `--list-checks` should help:

```shell
$ tmt lint --list-checks
Linters available for tests
C000: fmf node should pass the schema validation
C001: summary key should be set, it is very useful for quick inspection
C002: summary should not exceed 50 characters
T001: all keys are known
T002: test script must be defined
T003: test directory path must be absolute
T004: test directory path must exist
T005: relevancy has been obsoleted by adjust
T006: coverage has been obsoleted by link
T007: manual test path is not an actual path
T008: manual test should be valid markdown

Linters available for plans
C000: fmf node should pass the schema validation
C001: summary key should be set, it is very useful for quick inspection
C002: summary should not exceed 50 characters
P001: all keys are known
P002: execute step must be defined with "how"
P003: execute step methods must be known
P004: discover step methods must be known
P005: remote fmf ids must be valid

Linters available for stories
C000: fmf node should pass the schema validation
C001: summary key should be set, it is very useful for quick inspection
C002: summary should not exceed 50 characters
S001: all keys are known
S002: story key must be defined
$ 
```

File: tmt/options.py [hunk 4]
```diff
@@ -167,6 +168,37 @@
...
+        '--failed-only',
+        is_flag=True,
+        help='Display only tests/plans/stories that fail a check.'),
+    click.option(
+        '--outcome-only',
+        metavar='OUTCOME',
+        multiple=True,
+        type=click.Choice([
+            member.value for member in tmt.lint.LinterOutcome.__members__.values()]),
+        help='Display only checks with the given outcome.')
```
Comment (lukaszachy) [inline tmt/options.py:198]: Could we  show the list of possible OUTCOME values? Currently one has to dig it from error message printed after invalid input value.

**Review by lukaszachy — COMMENTED** [review]
Could  '--outcome-only' be improved? I did `tmt lint --outcome-only warn` and 
```
/stories/features/prepare/all

/stories/features/prepare/minimal
warn C002 summary should not exceed 50 characters

/stories/features/prepare/repository
```

Technically it works as expected - Shows only specified outcome. However each object name is printed as well so it is hard to read.
Could this be changed into 'print object name only it there is some check output as well?
Similar to '--failed-only' does

**Comment by lukaszachy** [comment]
How about `--disable-checks CHECK-ID` ? 

These changes look great overall. Very nice usability improvement.

**Comment by happz** [comment]
> Could '--outcome-only' be improved? I did `tmt lint --outcome-only warn` and
> 
> ```
> /stories/features/prepare/all
> 
> /stories/features/prepare/minimal
> warn C002 summary should not exceed 50 characters
> 
> /stories/features/prepare/repository
> ```
> 
> Technically it works as expected - Shows only specified outcome. However each object name is printed as well so it is hard to read. Could this be changed into 'print object name only it there is some check output as well? Similar to '--failed-only' does

Sure, makes sense. Done in ec6f8a8387fd31a41a5196f977c0989ab79ec6d2.

**Comment by happz** [comment]
> How about `--disable-checks CHECK-ID` ?
> 
> These changes look great overall. Very nice usability improvement.

Done in eb6ea5621fb454b870eea17748c76cbe76689ffe

File: tests/lint/test.sh (hunk 1 above)
Comment (lukaszachy) [reply tests/lint/test.sh:44]: So on brief look at lint report I was thinking: Why the same check is reported twice under different numbers with different results?

File: tmt/options.py (hunk 4 above)
Comment (happz) [reply tmt/options.py:215]: Done in 69a3c6f1:

```
  --outcome-only skip|pass|fail|warn|fixed
                                  Display only checks with the given outcome.
```

File: tmt/lint.py (hunk 2 above)
Comment (happz) [reply tmt/lint.py]: Fixed in 477d21e8

File: tests/lint/test.sh (hunk 1 above)
Comment (happz) [reply tests/lint/test.sh:44]: As discussed, C001 and C002 are now a single check, fixed in 8b4d9ce2:

```
/stories/features/report-result
warn C001 summary key is missing

/stories/features/prepare/all
pass C001 summary key is set and is reasonably long

/stories/features/prepare/minimal
warn C001 summary should not exceed 50 characters
```

File: tmt/lint.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+
+Classes spiced with ``Lintable`` define their sets of linters. Consider the
+following examples:
+
+.. code-block:: python
+
+   # Methods whose names start with ``lint_*`` prefix are considered linters.
+   def lint_path_exists(self) -> LinterReturn:
+       # A linter must have a docstring which is then used to generate documentation,
+       # e.g. when ``lint --list-checks`` is called. The docstring must being with
```
Comment (KwisatzHaderach) [inline tmt/lint.py:21]:
```suggestion
       # e.g. when ``lint --list-checks`` is called. The docstring must begin with
```

File: tmt/lint.py [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+        """
+
+        for name in dir(cls):
+            if not name.startswith('lint_'):
+                continue
+
+            cls.get_linter_registry().append(tmt.lint.Linter(getattr(cls, name)))
+
+    @classmethod
+    def resolve_enabled_linters(
```
Comment (KwisatzHaderach) [inline tmt/lint.py:184]: It would be great if we could store enabled, or better yet disabled, checks in some config, e.g. in .fmf/something
basically the same thing we are doing for flake, mypy and now ruff
completely fine to do in another patch

File: tmt/lint.py [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+                ]
+
+        return linters
+
+    def lint(
+        self,
+        enable_checks: Optional[List[str]] = None,
+        disable_checks: Optional[List[str]] = None,
+        enforce_checks: Optional[List[str]] = None
+            ) -> Tuple[bool, List[LinterRuling]]:
```
Comment (KwisatzHaderach) [inline tmt/lint.py:217]: this indentation looks weird, I'm surprised flake did not complain, IMO would look better on one level or flipped - 8 spaces for the args and 4 for the return type

File: tmt/lint.py [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+                    rulings.append((linter, outcome, outcome, message))
+
+        return valid, rulings
+
+    @classmethod
+    def format_linters(cls) -> str:
+        hints: List[str] = []
+
+        for linter in sorted(cls.get_linter_registry(), key=lambda x: x.id):
+            # Thsi has been already verified when registering the linter
```
Comment (KwisatzHaderach) [inline tmt/lint.py:259]:
```suggestion
            # This has been already verified when registering the linter
```

File: tmt/lint.py [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+            assert linter.callback.__doc__ is not None
+
+            hints.append(textwrap.dedent(linter.callback.__doc__).strip())
+
+        return '\n'.join(hints)
+
+
+def filter_allowed_checks(
+        rulings: Iterable[LinterRuling],
+        outcomes: Optional[List[LinterOutcome]] = None) -> Generator[LinterRuling, None, None]:
```
Comment (KwisatzHaderach) [inline tmt/lint.py:269]: you already started with return type on new line, it might look better to keep that style here as well
```suggestion
        outcomes: Optional[List[LinterOutcome]] = None
        ) -> Generator[LinterRuling, None, None]:
```

File: tmt/lint.py (hunk 5 above)
Comment (happz) [reply tmt/lint.py]: Fixed in a9a62baf.

File: tmt/lint.py (hunk 7 above)
Comment (happz) [reply tmt/lint.py]: Fixed in a9a62baf.

File: tmt/lint.py (hunk 8 above)
Comment (happz) [reply tmt/lint.py]: Fixed in a9a62baf.

**Review by KwisatzHaderach — APPROVED** [review]

File: tests/lint/test.sh (hunk 1 above)
Comment (lukaszachy) [reply tests/lint/test.sh]: thank you.

**Review by lukaszachy — APPROVED** [review]

**Comment by LecrisUT** [comment]
A note from #2008, should add a test condition for when the fmf root is in a different path. Maybe that should be made a test fixture/conftest.py so that it can be added to all test.

**Comment by happz** [comment]
A test for #2006 added in fa2520b0

**Comment by psss** [comment]
What about using a different color for the `skip` label? Currently it can be easily confused with `pass`.

**Comment by LecrisUT** [comment]
Green/Gray/Red/Yellow for pass/skip/fail/warning? These should be standard

**Comment by psss** [comment]
When an invalid key is present, the problem is reported twice:

```
warn C000 key "requires" not recognized by schema, and does not match "^extra-" pattern
fail T001 unknown key "requires" is used
```

Is this intentional? When two unknown keys are provided, duplication is gone:

```
warn C000 fmf node failed schema validation
warn C001 summary key is missing
fail T001 unknown key "hm" is used
fail T001 unknown key "requires" is used
```

Not sure why.

**Comment by psss** [comment]
> Green/Gray/Red/Yellow for pass/skip/fail/warning? These should be standard

Sounds good to me. Maybe `blue` could be also suitable for `skip` as an "info" color.

**Comment by happz** [comment]
> > Green/Gray/Red/Yellow for pass/skip/fail/warning? These should be standard
> 
> Sounds good to me. Maybe `blue` could be also suitable for `skip` as an "info" color.

Changed to `white` in 5b356c5a as Click does not offer `gray`. Let me know whether it should be `blue` instead...

**Comment by happz** [comment]
> When an invalid key is present, the problem is reported twice:
> 
> ```
> warn C000 key "requires" not recognized by schema, and does not match "^extra-" pattern
> fail T001 unknown key "requires" is used
> ```
> 
> Is this intentional? When two unknown keys are provided, duplication is gone:
> 
> ```
> warn C000 fmf node failed schema validation
> warn C001 summary key is missing
> fail T001 unknown key "hm" is used
> fail T001 unknown key "requires" is used
> ```
> 
> Not sure why.

I do know why, fixed in e2169ff1.

File: tests/usability/main.fmf [hunk 10]
```diff
@@ -1,2 +1,5 @@
 summary: Allow command abbreviation
 description: If user writes part of a subcommand, tmt should support to match it.
+
+requires: foo
+#someotherkey: foo
```
Comment (psss) [inline tests/usability/main.fmf:5]: Testing leftover?

Comment (happz) [reply tests/usability/main.fmf]: Correct, fixed in 20985c1b

**Comment by psss** [comment]
> Changed to `white` in [5b356c5](https://github.com/teemtee/tmt/commit/5b356c5aa4c2e2bd11c27a2134a71d2b6ff0ed21) as Click does not offer `gray`. Let me know whether it should be `blue` instead...

Thanks. Just tried with my light terminal and the `skip` tags are hardly readable :( Let's go with `blue` instead.

**Comment by happz** [comment]
> > Changed to `white` in [5b356c5](https://github.com/teemtee/tmt/commit/5b356c5aa4c2e2bd11c27a2134a71d2b6ff0ed21) as Click does not offer `gray`. Let me know whether it should be `blue` instead...
> 
> Thanks. Just tried with my light terminal and the `skip` tags are hardly readable :( Let's go with `blue` instead.

Changed to blue in 8fc1e89c

File: tmt/cli.py [hunk 11]
```diff
@@ -775,22 +875,36 @@ def plans_show(context: Context, **kwargs: Any) -> None:
 @click.pass_context
 @filter_options
 @fmf_source_options
+@lint_options
+@fix_options
```
Comment (psss) [inline tmt/cli.py:879]: Why are you adding `fix_options` here? We don't have any automated fixes for plans. Or do we? The same question for stories.

File: tmt/lint.py [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+                continue
+
+            cls.get_linter_registry().append(tmt.lint.Linter(getattr(cls, name)))
+
+    @classmethod
+    def resolve_enabled_linters(
+            cls,
+            enable_checks: Optional[List[str]] = None,
+            disable_checks: Optional[List[str]] = None
+            ) -> List[Linter]:
```
Comment (psss) [inline tmt/lint.py:188]: Please, add a docstring.

File: tmt/lint.py [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+                    else:
+                        rulings.append((linter, outcome, outcome, message))
+
+                else:
+                    rulings.append((linter, outcome, outcome, message))
+
+        return valid, rulings
+
+    @classmethod
+    def format_linters(cls) -> str:
```
Comment (psss) [inline tmt/lint.py:254]: Please, add a short docstring.

File: tmt/lint.py [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+
+def format_rulings(rulings: Iterable[LinterRuling]) -> Generator[str, None, None]:
+    """
+    Format rulings for printing or logging.
+
+    :param rulings: rulings to format.
+    :yields: rulings formatted as separate strings.
+    """
+
+    display_eventual = any(
```
Comment (psss) [inline tmt/lint.py:295]: Perhaps a short comment here explaining what `display_eventual` is would be handy.

File: tmt/base.py [hunk 15]
```diff
@@ -1503,106 +1664,120 @@ def show(self) -> None:
...
 
-    @staticmethod
-    def _lint_discover_fmf(discover: tmt.steps._RawStepData) -> bool:
-        """ Lint fmf discover method """
-        # Validate remote id and translate to human readable errors
-        fmf_id_data = cast(
-            _RawFmfId,
-            {key: value for key, value in discover.items() if key in ['url', 'ref', 'path']}
-            )
+        yield LinterOutcome.PASS, 'correct key are used'
```
Comment (psss) [inline tmt/base.py:1680]:
```suggestion
        yield LinterOutcome.PASS, 'correct keys are used'
```

File: tests/story/lint/test.sh [hunk 16]
```diff
@@ -9,25 +9,26 @@ rlJournalStart
 
     rlPhaseStartTest "Correct story"
         rlRun -s "tmt stories lint good" 0
-        rlAssertGrep "pass correct attributes are used" $rlRun_LOG
-        rlAssertNotGrep "warn summary" $rlRun_LOG
+        rlAssertGrep "pass C000 fmf node passes schema validation" $rlRun_LOG
+        rlAssertGrep "pass S001 correct key are used" $rlRun_LOG
```
Comment (psss) [inline tests/story/lint/test.sh:13]:
```suggestion
        rlAssertGrep "pass S001 correct keys are used" $rlRun_LOG
```

File: tests/story/lint/test.sh [hunk 17]
```diff
@@ -9,25 +9,26 @@ rlJournalStart
...
+        rlAssertGrep "warn C001 summary key is missing" $rlRun_LOG
+        rlAssertGrep "fail S001 unknown key \"exampleGG\" is used" $rlRun_LOG
 
         rlRun -s "tmt stories lint missing_story" 1
-        rlAssertGrep "fail story is required" $rlRun_LOG
-        rlAssertGrep "pass correct attributes are used" $rlRun_LOG
-        rlRun "rm $rlRun_LOG"
+        rlAssertGrep "warn C000 fmf node failed schema validation" $rlRun_LOG
+        rlAssertGrep "warn C001 summary key is missing" $rlRun_LOG
+        rlAssertGrep "pass S001 correct key are used" $rlRun_LOG
```
Comment (psss) [inline tests/story/lint/test.sh:30]:
```suggestion
        rlAssertGrep "pass S001 correct keys are used" $rlRun_LOG
```

File: tmt/base.py [hunk 18]
```diff
@@ -1946,30 +2128,33 @@ def coverage(self, code: bool, test: bool, docs: bool) -> Tuple[bool, bool, bool
...
         if invalid_keys:
             for key in invalid_keys:
-                verdict(False, f"unknown attribute '{key}' is used")
-        else:
-            verdict(True, "correct attributes are used")
+                yield LinterOutcome.FAIL, f'unknown key "{key}" is used'
+
+            return
+
+        yield LinterOutcome.PASS, 'correct key are used'
```
Comment (psss) [inline tmt/base.py:2143]:
```suggestion
        yield LinterOutcome.PASS, 'correct keys are used'
```

File: tmt/base.py [hunk 19]
```diff
@@ -539,6 +542,10 @@ def __str__(self) -> str:
         """ Node name """
         return self.name
 
+    @classmethod
+    def by_tree(cls: Type[T], tree: 'tmt.Tree') -> List[T]:
```
Comment (psss) [inline tmt/base.py:546]: Please, add a docstring.

File: tmt/base.py [hunk 20]
```diff
@@ -690,19 +697,111 @@ def _export(self, *, keys: Optional[List[str]] = None) -> tmt.export._RawExporte
...
+                yield LinterOutcome.WARN, \
+                    f'value of "{json_path.split(".")[-1]}" is not "{match.group(2)}"'
+
+            for error, _ in errors:
+                yield from detect_unallowed_properties(error)
+                yield from detect_unallowed_properties_with_pattern(error)
+                yield from detect_enum_violations(error)
+
+                # Validation errors can have "context", a list of "sub" errors encountered during
+                # validation. Interesting ones are identified &  added to our error message.
```
Comment (psss) [inline tmt/base.py:780]:
```suggestion
                # validation. Interesting ones are identified & added to our error message.
```

File: tmt/base.py [hunk 21]
```diff
@@ -972,88 +1082,139 @@ def show(self) -> None:
...
+        """ T004: test directory path must exist """
+
+        if not self.path:
+            yield LinterOutcome.FAIL, 'directory path is not set'
+            return
+
+        test_path = Path(self.node.root) / self.path.unrooted()
+
+        if not test_path.exists():
+            yield LinterOutcome.FAIL, f'test path "{test_path}" does not exist'
```
Comment (psss) [inline tmt/base.py:1132]:
```suggestion
            yield LinterOutcome.FAIL, f"test path '{test_path}' does not exist"
```

Let's consistently use [single quotes](https://tmt.readthedocs.io/en/latest/contribute.html#introduction) when reporting messages to the user.

File: tmt/base.py [hunk 22]
```diff
@@ -972,88 +1082,139 @@ def show(self) -> None:
...
-                verdict(False, f"unknown attribute '{key}' is used")
-        else:
-            verdict(True, "correct attributes are used")
 
-        # Check if the format of Markdown file respects the specification
-        # https://tmt.readthedocs.io/en/latest/spec/tests.html#manual
-        if self.manual:
-            valid &= self._lint_manual(test_path)
+        if coverage is None:
+            yield LinterOutcome.SKIP, 'legacy coverage not detected'
```
Comment (psss) [inline tmt/base.py:1166]:
```suggestion
            yield LinterOutcome.SKIP, "legacy 'coverage' field not detected"
```

File: tmt/base.py [hunk 23]
```diff
@@ -972,88 +1082,139 @@ def show(self) -> None:
...
+            return
 
-        return valid
+        yield LinterOutcome.FAIL, 'coverage has been obsoleted by link'
```
Comment (psss) [inline tmt/base.py:1169]:
```suggestion
        yield LinterOutcome.FAIL, "the 'coverage' field has been obsoleted by 'link'"
```

File: tmt/base.py [hunk 24]
```diff
@@ -1818,6 +1996,10 @@ def __init__(
                          raise_on_validation_error=raise_on_validation_error, **kwargs)
         self._update_metadata()
 
+    @classmethod
+    def by_tree(cls, tree: 'tmt.Tree') -> List['Story']:
```
Comment (psss) [inline tmt/base.py:2000]: Please, add a docstring.

**Review by psss — CHANGES_REQUESTED** [review]
This is a great enhancement! And will be superuseful. Quite an elegant implementation! Found some minor issues, a couple of questions, nothing serious.

File: tmt/cli.py (hunk 11 above)
Comment (happz) [reply tmt/cli.py:879]: I believe we don't have any - I felt it would be better to include the options to present the consistent set of options across all `lint` subcommands. If there are no automated fixes, those options would turn into no-op, but I don't see that as a problem. In my books, not having to remember which of 4 `lint` commands does not accept `--fix` is a good benefit, and once we add the first fix, it'd work out of the box, as we couldn't forget to add `--fix` option.

Comment (psss) [reply tmt/cli.py:879]: Ok, so it will be just prepared. Understand. Fine with me.

File: tmt/lint.py (hunk 12 above)
Comment (happz) [reply tmt/lint.py:188]: Added in 21c4e01c.

File: tmt/lint.py (hunk 13 above)
Comment (happz) [reply tmt/lint.py:281]: Added in 21c4e01c.

File: tmt/lint.py (hunk 14 above)
Comment (happz) [reply tmt/lint.py:335]: Added in 21c4e01c.

File: tmt/base.py (hunk 15 above)
Comment (happz) [reply tmt/base.py]: Fixed in 21c4e01c.

File: tests/story/lint/test.sh (hunk 16 above)
Comment (happz) [reply tests/story/lint/test.sh]: Fixed in 21c4e01c.

File: tests/story/lint/test.sh (hunk 17 above)
Comment (happz) [reply tests/story/lint/test.sh]: Fixed in 21c4e01c.

File: tmt/base.py (hunk 18 above)
Comment (happz) [reply tmt/base.py]: Fixed in 21c4e01c.

File: tmt/base.py (hunk 19 above)
Comment (happz) [reply tmt/base.py]: Added in 21c4e01c.

File: tmt/base.py (hunk 20 above)
Comment (happz) [reply tmt/base.py]: Fixed in 21c4e01c.

File: tmt/base.py (hunk 21 above)
Comment (happz) [reply tmt/base.py]: Fixed in 21c4e01c.

File: tmt/base.py (hunk 22 above)
Comment (happz) [reply tmt/base.py]: Fixed in 21c4e01c.

File: tmt/base.py (hunk 23 above)
Comment (happz) [reply tmt/base.py]: Fixed in 21c4e01c.

File: tmt/base.py (hunk 24 above)
Comment (happz) [reply tmt/base.py]: Added in 21c4e01c.

**Review by psss — APPROVED** [review]
Thanks for addressing all comments!

**Comment by psss** [comment]
The only failure is `/tests/finish/ansible` error on fetching the `centos:7` image. Happened multiple time recently. Filed as #2063.
